### PR TITLE
[Scan to update inventory] Part 3 – Support for manual quantity update

### PR DESCRIPTION
**Pull request by samiuelson**
This is **Part 3** of the Scan to update inventory project
Part 1 – https://github.com/woocommerce/woocommerce-android/pull/10250
Part 2 – https://github.com/woocommerce/woocommerce-android/pull/10253
Part 4 – https://github.com/woocommerce/woocommerce-android/pull/10313

This PR allows updating the product's stock quantity manually, by entering the stock quantity value with keyboard.

### Testing instructions

1. Open product list
2. Click "barcode scanner" icon
3. Scan barcode with the SKU of existing product (you can use online barcode generator like [this one](https://barcode.tec-it.com/en/?data=ACGADO))
4. Verify product is found and the bottom sheet is opened
5. Enter quantity manually and click the "update" button
6. Verify product update is successful

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/4527432/cc1706cd-a82f-43f0-bd8a-ad6767d54de9

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by kidinov** [comment]
https://github.com/woocommerce/woocommerce-android/assets/4923871/f73a4580-5b6d-4ee1-a3b2-4c42eca505e7

I am not sure how to improve that; maybe a longer delay after the barcode is recognized, or maybe a confirmation dialog? Currently hard to understand what's wrong

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheet.kt [hunk 1]
```diff
@@ -76,23 +92,62 @@ fun QuickInventoryUpdateBottomSheet(
...
+                        text = productState.product.quantity.toString(),
+                        selection = TextRange(productState.product.quantity.toString().length)
+                    ),
+                    onValueChange = {
+                        onManualQuantityEntered(it.text)
+                    },
+                    singleLine = true,
+                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
+                    textStyle = MaterialTheme.typography.body1.copy(
+                        textAlign = TextAlign.End,
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheet.kt]: There is something wrong with this TextField - I can not move the pointer 🤨


https://github.com/woocommerce/woocommerce-android/assets/4923871/69fe5dd1-a4d9-4680-9b3c-d600a18ea5d3

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheet.kt]: Also, I have doubts that it'll be clear for the users that this is editable field

**Comment by kidinov** [comment]
@samiuelson oh I see that in the design not stock manageable products will show another view. That'll be next PR?

image

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt [hunk 2]
```diff
@@ -1,21 +1,38 @@
...
-    private val args: QuickInventoryUpdateBottomSheetFragmentArgs by navArgs()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setStyle(DialogFragment.STYLE_NORMAL, R.style.Woo_Theme_BottomSheetDialog_WithKeyboard)
+    }
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        return super.onCreateDialog(savedInstanceState).apply {
+            if (this is BottomSheetDialog) {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt]: We may want to crash here if this for some reason is not `BottomSheetDialog`?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt [hunk 3]
```diff
@@ -1,21 +1,38 @@
...
-import androidx.navigation.fragment.navArgs
+import androidx.fragment.app.DialogFragment
+import com.google.android.material.bottomsheet.BottomSheetBehavior
+import com.google.android.material.bottomsheet.BottomSheetDialog
 import com.woocommerce.android.R
 import com.woocommerce.android.ui.compose.theme.WooTheme
 import com.woocommerce.android.viewmodel.fixedHiltNavGraphViewModels
 import com.woocommerce.android.widgets.WCBottomSheetDialogFragment
 
 class QuickInventoryUpdateBottomSheetFragment : WCBottomSheetDialogFragment() {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt]: Just curious, why did you decide to go with the BottomSheetDialog and shared ViewModel, not the bottom sheet "view" that will be part of `ScanToUpdateInventoryBarcodeScannerFragment`? It feels that "bottom sheet visible" is more like a state of the update inventory screen than a separate screen. You even have to cast to the state

`val productState = state.value as ScanToUpdateInventoryViewModel.ViewState.ProductLoaded`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/ScanToUpdateInventoryViewModel.kt [hunk 4]
```diff
@@ -119,7 +134,7 @@ class ScanToUpdateInventoryViewModel @Inject constructor(
     private suspend fun handleQuantityUpdateSuccess(oldProduct: Product, updatedProduct: Product) {
         val oldQuantity = oldProduct.stockQuantity
         val newQuantity = updatedProduct.stockQuantity
-        val quantityChangeString = "$oldQuantity ➡ $newQuantity"
+        val quantityChangeString = "${oldQuantity.toInt()} ➡ ${newQuantity.toInt()}"
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/ScanToUpdateInventoryViewModel.kt]: Do you know why it is double in the model? Maybe there are products, e.g. some kind of milk that can be, for instance 1.5L?

**Review by kidinov — COMMENTED** [review]
Thanks for working on this @samiuelson ! I left a few questions which probably worth considering before moving forward

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/ScanToUpdateInventoryViewModel.kt (hunk 4 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/ScanToUpdateInventoryViewModel.kt]: Good question —

TL;DR: Even though the DB column on backend is a "double" type the API converts any stock-quantity double values to integers (this is the default behavior in both admin and mobile app). Some 3rd party plugins override API's filtering and allow storing double values, but that's not supported by core currently.
 

> Core web handles decimal quantity input differently. Instead of rejecting non-integer stock quantities outright, it [accepts both integers and floats](https://href.li/?https://github.com/woocommerce/woocommerce/blob/0f4ccb836ed55031d8a0b7a7476b906461c2d2c4/includes/wc-formatting-functions.php#L470) and [uses a filter to return the integer value](https://href.li/?https://github.com/woocommerce/woocommerce/blob/b36a06ca080567bc6c863260267e62f65a2200d5/includes/wc-core-functions.php#L42) of the input. (Effectively, if you enter a decimal quantity on the web, once you update your product the decimal disappears and you’re left with an integer.) Third-party extensions like [WPC Product Quantity](https://href.li/?https://wordpress.org/plugins/wpc-product-quantity/) can replace that filter to allow non-integers values to be saved on the web.

Sources:
1. p91TBi-4iF-p2
2. https://github.com/woocommerce/woocommerce/issues/29664

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt (hunk 3 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt]: > Why not the bottom sheet "view" that will be part of ScanToUpdateInventoryBarcodeScannerFragment?

That would be ideal. I created a rough PoC, please take a look – https://github.com/woocommerce/woocommerce-android/pull/10296. There are a couple of problems:
- Since `ScanToUpdateInventoryBarcodeScannerFragment` extends the `BarcodeScanningFragment`, passing the bottomsheet's composable to the base class in order to inject it into the scanner composable makes the code quite complex and not maintainable.
- It doesn't help with casting the state, AFAICS

Let me know if you have idea how to show the bottom sheet as a view/composable from `ScanToUpdateInventoryBarcodeScannerFragment` directly in a cleaner way.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt (hunk 2 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt]: Changed here https://github.com/woocommerce/woocommerce-android/pull/10278/commits/066cd4b62880d17c891b3a61a8b9790df63daf67

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheet.kt (hunk 1 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheet.kt]: Enabled scrolling the cursor and migrated to OutlinedTextField here https://github.com/woocommerce/woocommerce-android/pull/10278/commits/e6b45694139ae39f98a7ad82a485ca8745533ed2

**Comment by samiuelson** [comment]
> I see that in the design not stock manageable products will show another view. That'll be next PR?

@kidinov, this design was added recently to the flow, we will discuss how to handle that case. At the moment we just show a snack bar. Anyway, it will come in another PR if we decide to go for it.

**Comment by samiuelson** [comment]
Thank you for the great review, @kidinov 🙇🏼 I addressed your findings, let me know what you think.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt (hunk 3 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/inventory/QuickInventoryUpdateBottomSheetFragment.kt]: @kidinov, I merged your barcode scanner improvements to my branches and managed to remove the redundant bottom sheet dialog fragment. I also got rid of inheriting from `BarcodeScannerFragment` – now the `ScanToUpdateInventoryBarcodeScannerFragment` reuses only the `BarcodeScanningViewModel` which I think is optimal.

**Review by kidinov — COMMENTED** [review]
@samiuelson thank you for all the improvements here! I noticed a few small bugs though

* You keep loading state in `handleProductNotFound`, so loading thing shown all the time

* The caret is black for some reason in the edit text field, not "the woo" color as it is in other places in the dark mode


* When I am trying to update quantity of the variable product I get an error:


Product:

```
Product(remoteId=1681, parentId=1661, name=Apple - L, black, description=, shortDescription=, slug=apple-l-black, type=variation, status=publish, catalogVisibility=visible, isFeatured=false, stockStatus=com.woocommerce.android.ui.products.ProductStockStatus$InStock@6cb96a7, backorderStatus=com.woocommerce.android.ui.products.ProductBackorderStatus$No@8c03754, dateCreated=Thu Dec 07 09:52:29 GMT+01:00 2023, firstImageUrl=https://site-for-woocommerce12a3fasdf45dfs6789.mystagingwebsite.com/wp-content/uploads/2023/02/product-1677159205-571152358.png, totalSales=0, reviewsAllowed=false, isVirtual=false, ratingCount=0, averageRating=0.0, permalink=https://site-for-woocommerce12a3fasdf45dfs6789.mystagingwebsite.com/product/apple/?attribute_pa_size=l&attribute_pa_color=black, externalUrl=, buttonText=, price=17.71, salePrice=null, regularPrice=17.71, taxClass=standard, isStockManaged=true, stockQuantity=151.0, sku=AP-20, shippingClass=, shippingClassId=0, isDownloadable=false, downloads=[], downloadLimit=-1, downloadExpiry=-1, purchaseNote=, numVariations=0, images=[Image(id=1660, name=Apple, source=https://site-for-woocommerce12a3fasdf45dfs6789.mystagingwebsite.com/wp-content/uploads/2023/02/product-1677159205-571152358.png, dateCreated=Thu Dec 07 09:52:29 GMT+01:00 2023)], attributes=[ProductAttribute(id=1, name=Size, terms=[], isVisible=true, isVariation=true), ProductAttribute(id=2, name=Color, terms=[], isVisible=true, isVariation=true)], saleEndDateGmt=null, saleStartDateGmt=null, isSoldIndividually=false, taxStatus=com.woocommerce.android.ui.products.ProductTaxStatus$Taxable@9ab56fd, isSaleScheduled=false, isPurchasable=true, menuOrder=0, categories=[], tags=[], groupedProductIds=[], crossSellProductIds=[], upsellProductIds=[], variationIds=[], length=0.0, width=0.0, height=0.0, weight=0.0, subscription=null, isSampleProduct=false, specialStockStatus=null, isConfigurable=false)
```

* The color of the scan icons is darker than it should be in the dark mode


* I would have some minimal width of the input field, because when it's empty looks weird



* It's possible to input spaces in this field. We may want to filter them

**Comment by samiuelson** [comment]
Thank you for the review @kidinov! I fixed the issues:

> * You keep loading state in `handleProductNotFound`, so loading thing shown all the time

Fixed – 0c18cbe573dc788833f8e957bee61efbabd38373 and improved – https://github.com/woocommerce/woocommerce-android/pull/10278/commits/9407ed70baae4b0fec43feaa4bc11cf8862b0019

> * The caret is black for some reason in the edit text field, not "the woo" color as it is in other places in the dark mode

Fixed – https://github.com/woocommerce/woocommerce-android/pull/10278/commits/aa3bb142fddf32cac6195de628413a50287d2cbe

> * When I am trying to update quantity of the variable product I get an error:

Good catch! The problem was the `getVariation` function – it was looking for variation in local DB only. I switched to `getVariationOrNull` which looks for variation in local DB and fetches it from the backend if not found in DB – https://github.com/woocommerce/woocommerce-android/pull/10278/commits/7fcac4dc1d93b9a35369493c6634061de41c989e

> * The color of the scan icons is darker than it should be in the dark mode

Fixed – https://github.com/woocommerce/woocommerce-android/pull/10278/commits/9108050f57856c17481b7ae97339d7db6bae2171

> * I would have some minimal width of the input field, because when it's empty looks weird

Added minWidth – https://github.com/woocommerce/woocommerce-android/pull/10278/commits/87d95cda21a6e962527ca14c6a12f00df57d08b4

> * It's possible to input spaces in this field. We may want to filter them

Allowing only positive Int or empty value now – https://github.com/woocommerce/woocommerce-android/pull/10278/commits/5779729cbd9ffc59d781ceef9de30d47566f7012

**Review by kidinov — APPROVED** [review]
LGTM!
